File: Sources/BraintreeThreeDSecure/BTThreeDSecureLookup.m
#import "BTThreeDSecureLookup_Internal.h"

// Swift Module Imports
#if __has_include(<Braintree/Braintree-Swift.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCore;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                            // Carthage
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif

@implementation BTThreeDSecureLookup

- (instancetype)initWithJSON:(BTJSON *)json {
    self = [super init];
    if (self) {
        _PAReq = [json[@"pareq"] asString];
        _MD = [json[@"md"] asString];
        _acsURL = [json[@"acsUrl"] asURL];
        _termURL = [json[@"termUrl"] asURL];
        _threeDSecureVersion = [json[@"threeDSecureVersion"] asString];
        _transactionID = [json[@"transactionId"] asString];
    }
    return self;
}

- (BOOL)requiresUserAuthentication {
    return self.acsURL != nil;
}

- (BOOL)isThreeDSecureVersion2 {
    return [self.threeDSecureVersion hasPrefix:@"2."];
}

@end

File: Sources/BraintreeThreeDSecure/BTThreeDSecureLookup.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

/// The result of a 3DS lookup.
/// Contains liability shift and challenge information.
@objcMembers public class BTThreeDSecureLookup: NSObject {

    // MARK: - Public Properties

    /// The "PAReq" or "Payment Authentication Request" is the encoded request message used to initiate authentication.
    public var paReq: String?

    // swiftlint:disable identifier_name
    /// The unique 3DS identifier assigned by Braintree to track the 3DS call as it progresses.
    public var md: String?
    // swiftlint:enable identifier_name

    ///  The URL which the customer will be redirected to for a 3DS Interface.
    ///  In 3DS 2, the presence of an acsURL indicates there is a challenge as it would otherwise frictionlessly complete without an acsURL.
    public var acsURL: URL?

    /// The termURL is the fully qualified URL that the customer will be redirected to once the authentication completes.
    public var termURL: URL?

    /// The full version string of the 3DS lookup result.
    public var threeDSecureVersion: String?

    /// Indicates a 3DS 2 lookup result.
    public var isThreeDSecureVersion2: Bool = false

    /// This a secondary unique 3DS identifier assigned by Braintree to track the 3DS call as it progresses.
    public var transactionID: String?

    /// Indicates that a 3DS challenge is required.
    public var requiresUserAuthentication: Bool = false

    // MARK: - Initializer

    init(json: BTJSON) {
        paReq = json["pareq"].asString()
        md = json["md"].asString()
        acsURL = json["acsUrl"].asURL()
        termURL = json["termUrl"].asURL()
        threeDSecureVersion = json["threeDSecureVersion"].asString()
        isThreeDSecureVersion2 = threeDSecureVersion?.hasPrefix("2.") ?? false
        transactionID = json["transactionId"].asString()
        requiresUserAuthentication = acsURL != nil
    }
}
